timestep.c

```diff
@@ -224,46 +255,63 @@ void advance_and_find_timesteps(void)
 
     if(badstepsizecount_global) {
         message(0, "bad timestep spotted: terminating and saving snapshot.\n");
-        All.NumCurrentTiStep = 0;
         savepositions(999999, 0);
         endrun(0, "Ending due to bad timestep");
     }
 
 
-    if(All.PM_Ti_endstep == All.Ti_Current)	/* need to do long-range kick */
+    if(is_PM_timestep(All.Ti_Current))
     {
-        advance_long_range_kick(All.MaxTiStepDisplacement);
+        /* Note this means we do a half kick on the first timestep.
+         * Which means we should also do a half-kick just before output.*/
+        const unsigned int tistart = get_kick_ti(PM_Ti.start, PM_Ti.step);
+        unsigned int tiend =  get_kick_ti(PM_Ti.start + PM_Ti.step, new_PM_Ti_step);
+        if(do_half_kick)
+            tiend = PM_Ti.start + PM_Ti.step;
+        /* Do long-range kick */
+        do_the_long_range_kick(tistart, tiend);
+        PM_Ti.start += PM_Ti.step;
+        PM_Ti.step = new_PM_Ti_step;
     }
 
     walltime_measure("/Timeline");
 }
 
-/*Advance a long-range timestep and do the desired kick.*/
-void advance_long_range_kick(int PM_Timestep)
+/* Just apply half a kick, for when
+ * we just wrote a snapshot with only half the kick applied.*/
+void
+apply_half_kick()
 {
-    int i;
-    int dti = TIMEBASE;
-    while(dti > PM_Timestep)
-        dti >>= 1;
-    /*Make it a little larger so it will go through the output time.*/
-    dti <<=1;
-
-    if(All.Ti_Current == TIMEBASE)	/* we here finish the last timestep. */
-        dti = 0;
-
-    const int tstart = (All.PM_Ti_begstep + All.PM_Ti_endstep) / 2;
-    const int tend = All.PM_Ti_endstep + dti / 2;
-
-    const double Fgravkick = get_gravkick_factor(tstart, tend);
-
-    All.PM_Ti_begstep = All.PM_Ti_endstep;
-    All.PM_Ti_endstep = All.PM_Ti_begstep + dti;
-
-    if(All.PM_Ti_kick != tstart) {
-        endrun(0, "PM kick time stamp mismatched\n");
+    int pa;
+    walltime_measure("/Misc");
+    /* Now assign new timesteps and kick */
+    #pragma omp parallel for
+    for(pa = 0; pa < NumActiveParticle; pa++)
+    {
+        const int i = ActiveParticle[pa];
+        int bin = P[i].TimeBin;
+        unsigned int dti = bin ? (1 << bin) : 0;
+        /* Start of step*/
+        unsigned int tistart = P[i].Ti_begstep;
+        /* Midpoint of step*/
+        unsigned int tiend = get_kick_ti(P[i].Ti_begstep, dti);
+        /*This only changes particle i, so is thread-safe.*/
+        do_the_short_range_kick(i, tistart, tiend);
     }
+    /*Always do a PM half-kick, because this should be called just after a PM step*/
```

Comment:
We probably want to have this in a separate function for PM kick. It could make the outer loop closer to the formula in the paper. I'll take a look at this.